src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseSpanBasedStringConcat.cs

```diff
@@ -0,0 +1,310 @@
+                                anySubstringInvocations = true;
+                            break;
+                        default:
+                            return false;
+                    }
+                }
+
+                return anySubstringInvocations;
+            }
+
+            bool IsAnyDirectOrConditionalSubstringInvocation(IOperation operation)
+            {
+                if (operation is IConditionalAccessOperation conditionallAccessOperation)
+                    operation = conditionallAccessOperation.WhenNotNull;
+
+                return operation is IInvocationOperation invocation && symbols.IsAnySubstringMethod(invocation.TargetMethod);
+            }
+        }
+
+        internal static ImmutableArray<IOperation> FlattenBinaryOperation(IBinaryOperation root)
+        {
+            var stack = new Stack<IBinaryOperation>();
+            var builder = ImmutableArray.CreateBuilder<IOperation>();
+            GoLeft(root);
+
+            while (stack.Count != 0)
+            {
+                var current = stack.Pop();
+
+                if (current.LeftOperand is not IBinaryOperation leftBinary || leftBinary.OperatorKind != root.OperatorKind)
+                {
+                    builder.Add(current.LeftOperand);
+                }
+
+                if (current.RightOperand is not IBinaryOperation rightBinary || rightBinary.OperatorKind != root.OperatorKind)
+                {
+                    builder.Add(current.RightOperand);
+                }
+                else
+                {
+                    GoLeft(rightBinary);
+                }
+            }
+
+            return builder.ToImmutable();
+
+            void GoLeft(IBinaryOperation operation)
+            {
+                IBinaryOperation? current = operation;
+                while (current is not null && current.OperatorKind == root.OperatorKind)
+                {
+                    stack.Push(current);
+                    current = current.LeftOperand as IBinaryOperation;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Remove the built in implicit conversion on operands to concat.
+        /// In VB, the conversion can be to either string or object.
+        /// In C#, the conversion is always to object.
+        /// </summary>
+        internal static IOperation WalkDownBuiltInImplicitConversionOnConcatOperand(IOperation operand)
+        {
+            if (operand is not IConversionOperation conversion)
+                return operand;
+            if (!conversion.IsImplicit || conversion.Conversion.IsUserDefined)
+                return conversion;
+
+            switch (conversion.Language)
+            {
+                case LanguageNames.CSharp when conversion.Type.SpecialType is SpecialType.System_Object:
+                case LanguageNames.VisualBasic when conversion.Type.SpecialType is SpecialType.System_Object or SpecialType.System_String:
```

Comment:
Please move all language specific logic to language specific layers instead of checking language names - you already have similar abstract methods and language specific operation walking code, so please move this out as well.

Reply:
Fixed